File: src_local/utils/markdown_processor.py
"""
📝 Processador Estrutural de Markdown.
Decompõe documentos em nós lógicos (headers, code blocks, content) para normalização PhD.
"""
import re
import logging
from src_local.utils.markdown_util import MarkdownUtil

logger = logging.getLogger(__name__)

class MarkdownStructureProcessor:
    """Processador de baixo nível para fluxos de linhas Markdown."""
    
    def __init__(self, raw_lines):
        self.raw = raw_lines
        self.res = []
        self.state = {"seen": {}, "in_cb": False}
        logger.debug(f"📝 [MarkdownProcessor] Iniciando processamento de {len(raw_lines)} linhas.")

    def process(self):
        for i, line in enumerate(self.raw):
            stripped = line.rstrip()
            if self._handle_block(stripped, line): continue
            
            if stripped.strip().startswith('#'):
                self._handle_header(stripped.strip(), i)
            elif self._should_append(stripped):
                self.res.append(stripped)
        return self.res

    def _handle_block(self, stripped, original):
        if stripped.strip().startswith('```'):
            self.state["in_cb"] = not self.state["in_cb"]
            self.res.append(original)
            return True
        if self.state["in_cb"]:
            self.res.append(original)
            return True
        return False

    def _handle_header(self, line, idx):
        if self.res and self.res[-1]: self.res.append('')
        h = MarkdownUtil.deduplicate_header(line, self.state["seen"])
        self.res.append(h)
        MarkdownUtil.apply_header_padding(self.res, self.raw, idx)

    def _should_append(self, stripped):
        return stripped.strip() or (self.res and self.res[-1])

File: src_local/utils/markdown_processor.ts
import { MarkdownUtil } from "./markdown_util.ts";
import winston from "winston";

const logger = winston.child({ module: "MarkdownProcessor" });

/**
 * 📝 Processador Estrutural de Markdown.
 * Decompõe documentos em nós lógicos (headers, code blocks, content) para normalização PhD.
 */
export class MarkdownStructureProcessor {
    private raw: string[];
    private res: string[] = [];
    private state = { seen: {} as Record<string, number>, inCb: false };

    constructor(rawLines: string[]) {
        this.raw = rawLines;
    }

    /**
     * Processa a estrutura do documento.
     */
    process(): string[] {
        logger.debug(`📝 [MarkdownProcessor] Iniciando processamento de ${this.raw.length} linhas.`);

        for (let i = 0; i < this.raw.length; i++) {
            const line = this.raw[i];
            if (line === undefined) continue;
            const stripped = line.trimEnd();

            // Gerenciamento de blocos de código
            if (this.handleBlock(stripped, line)) continue;

            // Processamento de Cabeçalhos
            if (stripped.trim().startsWith('#')) {
                this.handleHeader(stripped.trim(), i);
            } else if (this.shouldAppend(stripped)) {
                // Conteúdo regular ou espaçamento horizontal
                this.res.push(stripped);
            }
        }

        // MD041: Garante que não comece com linhas vazias (será tratado no sanitizer se necessário)
        return this.res;
    }

    private handleBlock(stripped: string, original: string): boolean {
        if (stripped.trim().startsWith('```')) {
            this.state.inCb = !this.state.inCb;
            this.res.push(original);
            return true;
        }
        if (this.state.inCb) {
            this.res.push(original);
            return true;
        }
        return false;
    }

    private handleHeader(line: string, idx: number): void {
        // MD022: Espaçamento acima
        if (this.res.length > 0 && this.res[this.res.length - 1] !== "") {
            this.res.push("");
        }

        const h = MarkdownUtil.deduplicateHeader(line, this.state.seen);
        this.res.push(h);

        // MD022: Espaçamento abaixo
        MarkdownUtil.applyHeaderPadding(this.res, this.raw, idx);
    }

    private shouldAppend(stripped: string): boolean {
        // Evita linhas em branco consecutivas (MD012)
        return stripped.trim() !== "" || (this.res.length > 0 && this.res[this.res.length - 1] !== "");
    }
}
